### PR TITLE
SYCL: Use sycl::ext::oneapi::experimental::auto_range to fix UB in RangePolicy parallel_for

### LINKED ISSUE #6745: UB when using desul's array lock on SYCL with RangePolicy
This test:

https://github.com/kokkos/kokkos/blob/2dc7cbcc9cebe83569a644ccf3a7c6e825e6ec4a/core/unit_test/TestAtomics.hpp#L537-L538

ultimately ends enqueued via SYCL's `range` version of `parallel_for` (vs. `nd_range` being an alternative).

desul's implementation at https://github.com/desul/desul/blob/fd6cd0639863a48ae0c57ed5db286955d6a412e2/atomics/include/desul/atomics/Lock_Based_Fetch_Op_SYCL.hpp#L35 uses https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_free_function_queries.asciidoc extension that specifies in https://github.com/intel/llvm/blob/f694c841e234835bc08578c905b7e904a1a57b4b/sycl/doc/extensions/experimental/sycl_ext_oneapi_free_function_queries.asciidoc?plain=1#L199-L209 that

> sub_group this_sub_group() ... shall only be called from within the `parallel_for` with `nd_range` as the argument
> If prerequisites are not fulfilled the behavior is undefined

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #6745. Some kernels are using `this_nd_item` which is only compatible with `sycl::nd_range/sycl::nd_item` but not `sycl::range/sycl::item`. For `RangePolicy` `parallel_for` we were using the former.
To avoid prescribing a particular work group size, we can leverage `sycl::ext::oneapi::experimental::auto_range` that lets the compiler choose something suitable (if available). If the extension is unavailable, this pull request doesn't change behavior.

### CODE REVIEW

File: core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp [hunk 1]
```diff
@@ -89,12 +94,33 @@ class Kokkos::Impl::ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>,
...
+            policy.begin(), functor, actual_range};
+        // Round the actual range up to the closest power of two not exceeding
+        // the maximum workgroup size
+        const auto max_wgroup_size =
+            q.get_device().get_info<sycl::info::device::max_work_group_size>();
+        const auto wgroup_size_multiple = Kokkos::bit_floor(
+            std::min<std::size_t>(max_wgroup_size, actual_range));
+
+        const auto launch_range = (actual_range + wgroup_size_multiple - 1) /
+                                  wgroup_size_multiple * wgroup_size_multiple;
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp:111]: So we can't rely on the implementation to adjust the size of the range?

**Review by dalg24 — COMMENTED** [review]
Please briefly explains how this resolves the UB that had been reported and introduce the experimental sycl feature you are leveraging.
https://github.com/intel/llvm/blob/cf6f876b339a7ea3b2c05fe5091134f328ef3814/sycl/doc/extensions/experimental/sycl_ext_oneapi_auto_local_range.asciidoc#sycl_ext_oneapi_auto_local_range

File: core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp:111]: The implementation will choose a workgroup size that divides the range which means 1 if the number of work items is a prime, see https://github.com/kokkos/kokkos/issues/6745#issuecomment-1912734891. We assume that the selected subgroup size will always be a power-of-two and so we increase the range accordingly and guard in the kernel for the actual range.

File: core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp [hunk 2]
```diff
@@ -37,14 +41,15 @@ struct FunctorWrapperRangePolicyParallelFor {
...
+#endif
 
 // Same as above but for a user-provided workgroup size
 template <typename FunctorWrapper, typename Policy>
 struct FunctorWrapperRangePolicyParallelForCustom {
   using WorkTag = typename Policy::work_tag;
 
-  void operator()(sycl::item<1> item) const {
-    const typename Policy::index_type id = item.get_linear_id();
+  void operator()(sycl::nd_item<1> item) const {
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp:51]: Yes, see https://github.com/intel/llvm/pull/11067.

**Review by nliber — APPROVED** [review]

File: core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp:51]: Please elaborate.  Are you referring to some issue that say that the dpc++ implementation allowed that conversion but, that the spec does not say that it is allowed, and that they recently corrected that oversight which forces us to fix that code?

[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp:51]: Yes, exactly, we could use `sycl::item` with `sycl::nd_range` before but newer compiler versions require `sycl::nd_item`.

**Comment by dalg24** [comment]
It is not obvious to me what part of this PR is keeping up with an implementation aligning with the spec and which is avoiding UB.  If the UB is mitigated only when the SYCL extension is detected, then please state it clearly.

**Comment by masterleinad** [comment]
> If the UB is mitigated only when the SYCL extension is detected, then please state it clearly.

Done.

**Review by dalg24 — APPROVED** [review]

**Comment by dalg24** [comment]
You want to rebase on develop to check that the CI passes.
The run is 5mo old. How confident are you that it is fine?

**Comment by masterleinad** [comment]
> You want to rebase on develop to check that the CI passes.

Sure. Let me rebase.

**Comment by dalg24** [comment]
It just occurred to me: do you want to include whether `SYCL_EXT_ONEAPI_AUTO_LOCAL_RANGE` was defined when you print configuration or just merge as is?

**Comment by masterleinad** [comment]
> It just occurred to me: do you want to include whether `SYCL_EXT_ONEAPI_AUTO_LOCAL_RANGE` was defined when you print configuration or just merge as is?

I think it's fine to merge as-is. I also confirmed that the incremental tests still work on the testbed.

**Pull request merged**